### PR TITLE
Add conf for specifying flint checkpoint location

### PR DESCRIPTION
**Pull request by noCharger**
### Description
Add conf for specifying a default checkpoint location to be re-used for index creation

### Issues Resolved
https://github.com/opensearch-project/opensearch-spark/issues/102

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: flint-spark-integration/src/main/scala/org/opensearch/flint/spark/FlintSparkIndexBuilder.scala [hunk 1]
```diff
@@ -139,4 +144,37 @@ abstract class FlintSparkIndexBuilder(flint: FlintSpark) {
...
+   * @param options
+   *   The original FlintSparkIndexOptions
+   * @return
+   *   Updated FlintSparkIndexOptions
+   */
+  private def updateOptionWithDefaultCheckpointLocation(
+      indexName: String,
+      options: FlintSparkIndexOptions): FlintSparkIndexOptions = {
+
+    val checkpointLocationRootDirOption = new FlintSparkConf(
```
Comment (vamsi-amazon) [inline flint-spark-integration/src/main/scala/org/opensearch/flint/spark/FlintSparkIndexBuilder.scala:162]: Add IT for this.

File: docs/index.md [hunk 2]
```diff
@@ -539,6 +539,7 @@ In the index mapping, the `_meta` and `properties`field stores meta and schema i
 - `spark.flint.optimizer.enabled`: default is true. enable the Flint optimizer for improving query performance.
 - `spark.flint.optimizer.covering.enabled`: default is true. enable the Flint covering index optimizer for improving query performance.
 - `spark.flint.index.hybridscan.enabled`: default is false.
+- `spark.flint.index.checkpointLocation.rootDir`: default is None.
```
Comment (vamsi-amazon) [inline docs/index.md:542]: Can you add description on the utility of this parameter.

File: flint-spark-integration/src/main/scala/org/opensearch/flint/spark/FlintSparkIndexBuilder.scala [hunk 3]
```diff
@@ -139,4 +144,37 @@ abstract class FlintSparkIndexBuilder(flint: FlintSpark) {
...
+      Collections.emptyMap[String, String]).checkpointLocationRootDir
+
+    if (options.checkpointLocation().isEmpty) {
+      checkpointLocationRootDirOption match {
+        case Some(checkpointLocationRootDir) =>
+          // Currently, deleting and recreating the flint index will enter same checkpoint dir.
+          // Use a UUID to isolate checkpoint data.
+          val checkpointLocation =
+            s"${checkpointLocationRootDir.stripSuffix("/")}/$indexName/${UUID.randomUUID().toString}"
```
Comment (vamsi-amazon) [inline flint-spark-integration/src/main/scala/org/opensearch/flint/spark/FlintSparkIndexBuilder.scala:171]: Have we decided to go with only untill UUID, no "checkpoint" after that?

File: integ-test/src/integration/scala/org/opensearch/flint/spark/FlintSparkCoveringIndexITSuite.scala [hunk 4]
```diff
@@ -117,6 +118,47 @@ class FlintSparkCoveringIndexITSuite extends FlintSparkSuite {
...
+
+      val indexData = flint.queryIndex(testFlintIndex)
+      checkAnswer(indexData, Seq(Row("Hello", 30), Row("World", 25)))
+
+      val index = flint.describeIndex(testFlintIndex)
+      index shouldBe defined
+
+      val checkpointLocation = index.get.options.checkpointLocation()
+      assert(checkpointLocation.isDefined, "Checkpoint location should be defined")
+      assert(
```
Comment (vamsi-amazon) [inline integ-test/src/integration/scala/org/opensearch/flint/spark/FlintSparkCoveringIndexITSuite.scala:156]: is there an issue with testing exact location?

Comment (noCharger) [reply integ-test/src/integration/scala/org/opensearch/flint/spark/FlintSparkCoveringIndexITSuite.scala:156]: The UUID is random, same as the withTempDir call generated checkpointDir

File: flint-spark-integration/src/main/scala/org/opensearch/flint/spark/FlintSparkIndexBuilder.scala (hunk 3 above)
Comment (noCharger) [reply flint-spark-integration/src/main/scala/org/opensearch/flint/spark/FlintSparkIndexBuilder.scala:171]: This directory should be used for checkpoint only, we can extend it in the future if there's additional use case

**Review by vamsi-amazon — APPROVED** [review]

**Pull request merged**